### PR TITLE
Assert coercion behavior with Hive

### PR DESCRIPTION
**Pull request by phd3**
Based on the suggestion in https://github.com/prestosql/presto/pull/5575#issuecomment-713427731, this PR adds assertions for coercions in different versions of Hive in addition to Presto.

### CODE REVIEW

**Comment by phd3** [comment]
@findepi do the flaky tests not get retried with product tests? The failure below is marked flaky in https://github.com/prestosql/presto/pull/5799

```
2020-11-09T16:35:50.8547284Z tests               | 2020-11-09 22:20:50 INFO: not retrying; FlakyTestRetryAnalyzer disabled
2020-11-09T16:35:50.8874482Z tests               | 2020-11-09 22:20:50 INFO: FAILURE     /    io.prestosql.tests.hive.TestHiveTransactionalTable.testFilesForAbortedTransactionsIgnored (Groups: hive_transactional) took 15.1 seconds
2020-11-09T16:35:50.8883264Z tests               | 2020-11-09 22:20:50 SEVERE: Failure cause:
2020-11-09T16:35:50.8883759Z tests               | java.lang.AssertionError: Expected row count to be <4>, but was <6>; rows=[[1], [1], [1], [2], [2], [2]]
2020-11-09T16:35:50.8885960Z tests               | 	at io.prestosql.tests.hive.TestHiveTransactionalTable.testFilesForAbortedTransactionsIgnored(TestHiveTransactionalTable.java:892)
2020-11-09T16:35:50.8909688Z tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2020-11-09T16:35:50.8911341Z tests               | 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2020-11-09T16:35:50.8913487Z tests               | 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2020-11-09T16:35:50.8915118Z tests               | 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2020-11-09T16:35:50.8916604Z tests               | 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
2020-11-09T16:35:50.8918216Z tests               | 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
2020-11-09T16:35:50.8919410Z tests               | 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
2020-11-09T16:35:50.8920969Z tests               | 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
2020-11-09T16:35:50.8922757Z tests               | 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
2020-11-09T16:35:50.8924608Z tests               | 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
2020-11-09T16:35:50.8926364Z tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-11-09T16:35:50.8928209Z tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-11-09T16:35:50.8929393Z tests               | 	at java.base/java.lang.Thread.run(Thread.java:834)
```

May be we need to set `io.prestosql.testng.services.FlakyTestRetryAnalyzer.enabled` for product tests as well?

**Comment by findepi** [comment]
> @findepi do the flaky tests not get retried with product tests? The failure below is marked flaky in #5799

Should be, since https://github.com/prestosql/presto/commit/41b5f377a1, but maybe i did something wrong.

Note however, that `@Flaky` are not retried when run locally.
Rationale is: typically when a flaky test is reproducible locally, it's also quite easy to fix, so no point in retrying and hiding the failure.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/TestRun.java [hunk 1]
```diff
@@ -262,7 +262,8 @@ private Environment getEnvironment()
                                         "-XX:MinHeapFreeRatio=10",
                                         "-XX:MaxHeapFreeRatio=10",
                                         "-Djava.util.logging.config.file=/docker/presto-product-tests/conf/tempto/logging.properties",
-                                        "-Duser.timezone=Asia/Kathmandu")
+                                        "-Duser.timezone=Asia/Kathmandu",
+                                        "-Dio.prestosql.testng.services.FlakyTestRetryAnalyzer.enabled=true")
```
Comment (findepi) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/cli/TestRun.java:266]: remember to drop this

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCoercion.java [hunk 2]
```diff
@@ -140,6 +146,24 @@
                         "STORED AS AVRO");
     }
 
+    enum ExecutorType
+    {
+        PRESTO(onPresto()),
+        HIVE(onHive());
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCoercion.java:152]: `onPresto()` could return something that's scoped for a test method or class.
(maybe it does already?)

thus do it lazy, only when `getExecutor` is called.
or, do like here
https://github.com/prestosql/presto/blob/34504af3ce2a4dfff4fecdb1d26d840787345220/presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java#L793-L797
where the enum is dumb

**Comment by phd3** [comment]
An unrelated failure: https://github.com/prestosql/presto/issues/2929

**Comment by phd3** [comment]
@findepi gentle ping :)

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCoercion.java [hunk 3]
```diff
@@ -561,4 +730,81 @@ private static Row project(Row row, int... columns)
...
+        {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            ColumnContext that = (ColumnContext) o;
+            return Objects.equals(hiveVersion, that.hiveVersion) && Objects.equals(format, that.format) && Objects.equals(
+                    column, that.column);
```
Comment (findepi) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCoercion.java:778]: fmt

**Review by findepi — APPROVED** [review]
(skimmed)

**Comment by phd3** [comment]
Merged #5884 - with an offline consent from @findepi.

**Pull request merged**